x-pack/plugins/reporting/server/export_types/printable_pdf/create_job/compatibility_shim.ts

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { KibanaRequest } from 'kibana/server';
+import { url as urlUtils } from '../../../../../../../src/plugins/kibana_utils/server';
+import type { LevelLogger } from '../../../lib';
+import { CreateJobFn, ReportingRequestHandlerContext } from '../../../types';
+import { validateUrls } from '../../common';
+import type { BaseParamsPDF, JobParamsPDF, JobParamsPDFLegacy, TaskPayloadPDF } from '../types';
+
+function isNewParams(jobParams: JobParamsPDF | JobParamsPDFLegacy): jobParams is JobParamsPDF {
+  return (jobParams as BaseParamsPDF).relativeUrls != null;
+}
+
+const getSavedObjectTitle = async (
+  objectType: string,
+  savedObjectId: string,
+  savedObjectsClient: any
+) => {
+  const savedObject = await savedObjectsClient.get(objectType, savedObjectId);
+  return savedObject.attributes.title;
+};
+
+const getSavedObjectRelativeUrl = (
+  objectType: string,
+  savedObjectId: string,
+  queryString: string
+) => {
+  const appPrefixes: Record<string, string> = {
+    dashboard: '/dashboard/',
+    visualization: '/visualize/edit/',
+    search: '/discover/',
+  };
+
+  const appPrefix = appPrefixes[objectType];
+  if (!appPrefix) throw new Error('Unexpected app type: ' + objectType);
+
+  const hash = appPrefix + urlUtils.encodeUriQuery(savedObjectId, true);
+
+  return `/app/kibana#${hash}?${queryString || ''}`;
+};
+
+/*
+ * The compatibility shim is responsible for migrating an older shape of the
+ * PDF Job Params into a newer shape, by deriving a report title and relative
+ * URL from a savedObjectId and queryString.
+ */
+export function compatibilityShim(
```

Comment:
These changes remove the "Factory" abstraction that was in the JS version of this file. That aligns with https://github.com/elastic/kibana/pull/106940